strapi/strapi
# fix(security-middleware): fix config merging with defaults

### What does it do?

Fixes merging logic in security middleware configuration.

### Why is it needed?

Currently, security middleware ignores user custom CSP rules for the directives `script-src/connect-src`, which blocks loading any custom scripts.

### How to test it?

1. Set custom script-src CSP rules in security middleware configuration `config/middleware`. Example:

```js
// config/middleware.ts
export default [
  'strapi::logger',
  'strapi::errors',
  {
    name: 'strapi::security',
    config: {
      contentSecurityPolicy: {
        useDefaults: true,
        directives: {
          'connect-src': ["'self'", 'https:'],
          'script-src': ["'self'", 'https://cdn.jsdelivr.net', 'https://unpkg.com'],
          upgradeInsecureRequests: null,
        },
      },
    },
  },
  ...
];
```

2. Start and request admin page from strapi server, and verify that custom CSP rules are respected.

```sh
curl -v https://localhost:1337/admin 2>&1 | grep content-security-policy
```

### Related issue(s)/PR(s)

This fixes #19589.
